libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java

```diff
@@ -18,6 +18,10 @@
 public abstract class AbstractRefCounted implements RefCounted {
     public static final String ALREADY_CLOSED_MESSAGE = "already closed, can't increment ref count";
 
+    public static boolean incrementIfPositive(AtomicInteger counter) {
```

Comment:
I kinda put this here because I figured that we were in-fact doing ref-counting in `KeyedLock` and didn't feel like adding a new utility class so it felt ok? In the end, `KeyLock` could just as well be an `AbstractRefCounted` with a `ReentrantLock` field in it I think :)

Reply:
Do you prefer this over making `AbstractRefCounted` use a `volatile int` field and a `VarHandle`? Presumably the lambda gets inlined nicely, but (as well as the extra indirection) note that this always has memory barriers as if doing the get and the set, even if zero, whereas the previous implementation skips the set barrier once closed.

Reply:
LGTMd the other PR will inline this one in the remaining spot then once that's merged.

Reply:
and it's gone and inlined :)